userspace/engine/rule_loader_compiler.cpp

```diff
@@ -444,10 +452,12 @@ void rule_loader::compiler::compile_rule_infos(
 		// This also compiles the filter, and might throw a
 		// falco_exception with details on the compilation
 		// failure.
+		sinsp_filter_compiler compiler(cfg.sources.at(r.source)->filter_factory, ast.get());
 		try {
-			source->ruleset->add(*out.at(rule_id), ast);
+			shared_ptr<gen_event_filter> filter(compiler.compile());
```

Comment:
Of this, I don't like the fact that we bundle a `sinsp_filter_compiler` inside the `rule_loader::compiler` class. In the future, we potentially may want to have multiple implementation of filters, and in general I don't think this should be part of this class' responsibilities. 

Given this, I propose two options. I'm leaning towards option 1 because it makes these changes less impactful and better modularizes each class' responsibilities.

**Option 1**
The only reason why we are passing a `gen_event_filter`, and compiling the filter here instead then in `filter_ruleset`, is to have access to the compiler's filter position in case of error. As such, we could simply create a new exception type (e.g. `falco_rule_compile_exception`) child of `falco_exception`. This new exception type would carry the AST position information in case of error. Then, make `filter_ruleset::add` throw both `falco_rule_compile_exception` and `falco_exception` by documentation, and catch those two errors here.

So here we would do a catch chain such as: if we have a `falco_rule_compile_exception`, then throw an error with a filter-condition-specific context with a specific position, otherwise if we have a generic `falco_exception` just throw an error with `r.cond_ctx`.

In this way, we would have all the information we need for information reporting here were we need it, by also making all these changes less impactful and by not changing the `filter_ruleset` interface (if not for the "throws" guarantees). Also, the `rule_loader::compile` class will not have more responsibilities than we should (e.g. compiling a filter), and we have a clear separation between the rule condition AST and its black-box compiled filter (of which each `filter_ruleset` implementation is solely responsible).



**Option 2**
Let's proceed with these changes, but in future PRs (reminder for ourselves):
- Make `gen_event_filter` become a real interface, so that there can actually be multiple implementations of it
- Add an intermediary class between `rule_loader::compiler` and `filter_ruleset` that takes care of 1) compiling the filter and report condition errors, 2) add it to the ruleset and to the state's rule list, and 3) enable/disable it on the ruleset. All these are becoming too many responsibilities for `rule_loader::compiler`, which is only supposed to resolve all the engine definitions in a list of rules.

Reply:
I think of each step of rule loading, compilation, and filter management being separate responsibilities, so I'm not sure moving the compilation into the filter_ruleset is a good idea either. To me the filter_ruleset is just about managing collections of filters, where it can optimize/prefilter/etc based on stuff like event type, as the current one does.

I guess I would think of compilation as something that could be plugged into the rule loader and accessed via event source, as we currently do with filter_factory/formatter_factory. In a fully aligned design, the engine would just be configured with compilers, and a compiler could in turn be configured with fields, so a filter factory would be responsible for compilation too.

So I guess I'd lean towards option 2 with some tbd on the actual interfaces.